internal/middleware/compression.go

```diff
@@ -18,27 +18,38 @@ import (
 	"bytes"
 	"compress/gzip"
 	"compress/zlib"
+	"fmt"
 	"io"
 	"net/http"
+
+	"github.com/golang/snappy"
+	"github.com/klauspost/compress/zstd"
 )
 
 const (
 	headerContentEncoding = "Content-Encoding"
-	headerValueGZIP       = "gzip"
+	CompressionGzip       = "gzip"
+	CompressionZlib       = "zlib"
+	CompressionDeflate    = "deflate"
+	CompressionSnappy     = "snappy"
+	CompressionZstd       = "zstd"
 )
 
-type CompressRoundTripper struct {
-	http.RoundTripper
-}
+type (
+	CompressRoundTripper struct {
+		RoundTripper    http.RoundTripper
+		CompressionType string
+	}
+)
 
-func NewCompressRoundTripper(rt http.RoundTripper) *CompressRoundTripper {
+func NewCompressRoundTripper(rt http.RoundTripper, compressionType string) *CompressRoundTripper {
 	return &CompressRoundTripper{
-		RoundTripper: rt,
+		RoundTripper:    rt,
+		CompressionType: compressionType,
```

Comment:
You can also store just a reference to the writer function. Like:

```golang
type CompressRoundTripper struct {
	RoundTripper    http.RoundTripper
	CompressionType string
	Writer          func(*bytes.Buffer) (io.WriteCloser, error)
}
```

And then:
```golang
func NewCompressRoundTripper(rt http.RoundTripper, compressionType string) *CompressRoundTripper {
	return &CompressRoundTripper{
		RoundTripper:    rt,
		CompressionType: compressionType,
		Writer:          writerFactory(compressionType),
	}
}
```

And then:

```golang
func writerFactory(compressionType string) func(*bytes.Buffer) (io.WriteCloser, error) {
	switch compressionType {
	case CompressionGzip:
		return func(buf *bytes.Buffer) (io.WriteCloser, error) {
			return gzip.NewWriter(buf), nil
		}
	default:
		return nil
	}
}
```

Reply:
Note: you probably want to have the call to `writerFactory` to happen in a separate statement and check for nil, indicating that the compression type is not known.

Reply:
done. Thanks for letting me know.